lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -25,533 +26,472 @@
+      auto addressFieldId = type.getFieldID((unsigned)ReadPortSubfield::addr);
+      if (ref.getFieldID() == enableFieldId ||
+          ref.getFieldID() == addressFieldId) {
+        for (auto dataField : fieldToVals[Node(memPort, dataFieldId)])
+          children.push_back(dataField);
+      }
+    }
+  }
+
+  void reportLoopFound(Value childVal, VisitingSet visiting) {
+    auto getName = [&](Node node) {
+      if (isa<SubfieldOp, SubindexOp, SubaccessOp>(node.getDefiningOp())) {
+        assert(!valRefersTo[node.getValue()].empty());
+        return getFieldName(*valRefersTo[node.getValue()].begin()).first;
       }
+      return getFieldName(node).first;
+    };
+    Node childNode(childVal, 0);
+    auto lastSignalName = getName(childNode);
+    auto errorDiag = mlir::emitError(
+        module.getLoc(),
+        "detected combinational cycle in a FIRRTL module, sample path: ");
+    if (!lastSignalName.empty())
+      errorDiag << module.getName() << "." << lastSignalName << " <- ";
+    visiting.popUntillVal(childVal, [&](Value visitingVal) {
+      auto signalName = getName(Node(visitingVal, 0));
+      if (!signalName.empty())
+        errorDiag << module.getName() << "." << signalName << " <- ";
     });
+    if (!lastSignalName.empty())
+      errorDiag << module.getName() << "." << lastSignalName;
+  }
+
+  LogicalResult handleConnects(Value dst, SmallVector<Node> &inputArgFields) {
 
-    auto traverseDFSFrom = [&](Value root, Node inputArg) -> LogicalResult {
-      if (visited.contains(root))
-        return success();
-      return dfsFromNode(root, inputArg);
+    bool onlyFieldZero = true;
+    auto pathsToOutPort = [&](Node dstNode) {
+      if (dstNode.getFieldID() != 0)
+        onlyFieldZero = false;
+      if (!dstNode.getValue().isa<BlockArgument>()) {
+        return failure();
+      }
+      onlyFieldZero = false;
+      for (auto inArg : inputArgFields) {
+        portPaths[inArg].insert(dstNode);
+      }
+      return success();
     };
-    for (auto root : rootValues) {
-      if (root.second.isValid())
-        visited.clear();
-      if (traverseDFSFrom(root.first, root.second).failed())
+    forallRefersTo(dst, pathsToOutPort);
+
+    if (onlyFieldZero) {
+      if (isa<RegOp, RegResetOp, SubfieldOp, SubaccessOp, SubindexOp>(
+              dst.getDefiningOp()))
         return failure();
     }
-
-    LLVM_DEBUG({
-      for (const auto &i1 : portPaths)
-        for (const auto &i2 : i1.second)
-          for (const auto &i3 : i2.second)
-            llvm::dbgs() << "\n node :" << i1.first << "," << i2.first
-                         << " is connected to " << i3.val << "," << i3.fieldId;
-    });
     return success();
   }
 
-  void printError(Value node, ValueOpPair &childNode) {
-    if (!printCycle || currentPath.back() == node) {
-      printCycle = false;
+  void setValRefsTo(Value val, Node ref) {
+    if (!val || !ref)
       return;
```

Comment:
Could `val` or `ref` be nullable? If it is an error, can we make this an assertion?